airflow/models/dag.py

```diff
@@ -2738,22 +2738,48 @@ def add_logger_if_needed(ti: TaskInstance):
         )
 
         tasks = self.task_dict
+
+        task_try_numbers: dict[tuple[str, int], int] = collections.defaultdict(int)
+
         self.log.debug("starting dagrun")
         # Instead of starting a scheduler, we run the minimal loop possible to check
         # for task readiness and dependency management. This is notably faster
         # than creating a BackfillJob and allows us to surface logs to the user
         while dr.state == DagRunState.RUNNING:
             schedulable_tis, _ = dr.update_state(session=session)
-            try:
-                for ti in schedulable_tis:
+
+            for ti in schedulable_tis:
+                try:
                     add_logger_if_needed(ti)
                     ti.task = tasks[ti.task_id]
                     _run_task(ti, session=session)
-            except Exception:
-                self.log.info(
-                    "Task failed. DAG will continue to run until finished and be marked as failed.",
-                    exc_info=True,
-                )
+                except Exception:
+                    if ti.state == TaskInstanceState.UP_FOR_RETRY:
+                        try_number = task_try_numbers[ti.task_id, ti.map_index]
+                        if try_number > ti.max_tries:
+                            ti.set_state(TaskInstanceState.FAILED)
+                        else:
+                            task_try_numbers[ti.task_id, ti.map_index] = try_number + 1
+                    self.log.info(
+                        "Task failed. DAG will continue to run until finished and be marked as failed.",
+                        exc_info=True,
+                    )
+            for ti in dr.get_task_instances(session=session, state=TaskInstanceState.SCHEDULED):
```

Comment:
This won't handle task instances in the deferred state though. I may not fully understand what this PR is fixing, but to me it seems like any deferred operator/sensor hangs on `dag.test()`.

Reply:
This additiona loop allows `test()` to wait for tasks in the deferred state. When they are resumed (by the triggerer), they will be picked up by this loop and executed. The function will appear hanging while the tasks are still in the deferred state because it’s waiting for them to end.

Reply:
Is the triggerer intended to be a available to `dag.test()' though? It definitely wouldn't be running in unit tests. My understanding is that this method is a crude replacement for scheduling so they you don't have to run a scheduler. Wouldn't the same idea apply to an external triggerer?

Reply:
Depends on how you view things I guess. The context of all this is pretty undefined to begin with. I would not object if someone wants to expand the scope to include the triggering part. It would be quite more involved than this PR though.

Reply:
Thanks. Yeah, I'm definitely unclear on the use of `dag.test()`. It's used in tests where (I'm assuming) an external triggerer isn't used. This may become more clear when the included test in this PR uses `DateTimeSensorAsync`.